Source/Core/Core/IOS/ES/ES.cpp

```diff
@@ -249,981 +249,1057 @@ u32 ES::OpenTitleContent(u32 CFD, u64 TitleID, u16 Index)
 
-    // Try to find the title key from a pre-installed ticket.
-    std::vector<u8> ticket = DiscIO::FindSignedTicket(m_addtitle_tmd.GetTitleId());
-    if (ticket.size() == 0)
-    {
-      return GetDefaultReply(ES_NO_TICKET_INSTALLED);
-    }
+  u32 content_fd = Memory::Read_U32(request.in_vectors[0].address);
+  INFO_LOG(IOS_ES, "IOCTL_ES_ADDCONTENTDATA: content fd %08x, "
+                   "size %d",
+           content_fd, request.in_vectors[1].size);
 
-    mbedtls_aes_context aes_ctx;
-    mbedtls_aes_setkey_dec(&aes_ctx, DiscIO::GetKeyFromTicket(ticket).data(), 128);
+  u8* data_start = Memory::GetPointer(request.in_vectors[1].address);
+  u8* data_end = data_start + request.in_vectors[1].size;
+  m_addtitle_content_buffer.insert(m_addtitle_content_buffer.end(), data_start, data_end);
+  return GetDefaultReply(IPC_SUCCESS);
+}
 
-    // The IV for title content decryption is the lower two bytes of the
-    // content index, zero extended.
-    TMDReader::Content content_info;
-    if (!m_addtitle_tmd.FindContentById(m_addtitle_content_id, &content_info))
-    {
-      return GetDefaultReply(ES_INVALID_TMD);
-    }
-    u8 iv[16] = {0};
-    iv[0] = (content_info.index >> 8) & 0xFF;
-    iv[1] = content_info.index & 0xFF;
-    std::vector<u8> decrypted_data(m_addtitle_content_buffer.size());
-    mbedtls_aes_crypt_cbc(&aes_ctx, MBEDTLS_AES_DECRYPT, m_addtitle_content_buffer.size(), iv,
-                          m_addtitle_content_buffer.data(), decrypted_data.data());
-
-    std::string path = StringFromFormat(
-        "%s%08x.app",
-        Common::GetTitleContentPath(m_addtitle_tmd.GetTitleId(), Common::FROM_SESSION_ROOT).c_str(),
-        m_addtitle_content_id);
-
-    File::IOFile fp(path, "wb");
-    fp.WriteBytes(decrypted_data.data(), decrypted_data.size());
-
-    m_addtitle_content_id = 0xFFFFFFFF;
-    break;
+IPCCommandResult ES::AddContentFinish(const IOCtlVRequest& request)
+{
+  _dbg_assert_msg_(IOS_ES, request.in_vectors.size() == 1,
+                   "IOCTL_ES_ADDCONTENTFINISH wrong number of inputs");
+
+  u32 content_fd = Memory::Read_U32(request.in_vectors[0].address);
+  INFO_LOG(IOS_ES, "IOCTL_ES_ADDCONTENTFINISH: content fd %08x", content_fd);
+
+  // Try to find the title key from a pre-installed ticket.
+  std::vector<u8> ticket = DiscIO::FindSignedTicket(m_addtitle_tmd.GetTitleId());
+  if (ticket.size() == 0)
+  {
+    return GetDefaultReply(ES_NO_TICKET_INSTALLED);
   }
 
-  case IOCTL_ES_ADDTITLEFINISH:
+  mbedtls_aes_context aes_ctx;
+  mbedtls_aes_setkey_dec(&aes_ctx, DiscIO::GetKeyFromTicket(ticket).data(), 128);
+
+  // The IV for title content decryption is the lower two bytes of the
+  // content index, zero extended.
+  TMDReader::Content content_info;
+  if (!m_addtitle_tmd.FindContentById(m_addtitle_content_id, &content_info))
   {
-    INFO_LOG(IOS_ES, "IOCTL_ES_ADDTITLEFINISH");
-    break;
+    return GetDefaultReply(ES_INVALID_TMD);
   }
+  u8 iv[16] = {0};
```

Comment:
Heh I can't believe you copied this @lioncash :P
u8 iv[16] = {
  (content_info.index >> 8) & 0xFF,
  content_info.index & 0xFF
}
???